### ISSUE TITLE
ganeti.cabal: static vs. generated

### ISSUE THREAD

**Issue by thielema** [ganeti/ganeti#1747]
Currently `ganeti.cabal` is constructed from `cabal/ganeti.template.cabal` by `cabal/cabal-from-modules.py`. But for construction `make` seems to already need to run Haskell code, using the incomplete `cabal/ganeti.template.cabal`. In the end the Haskell code seems to be build without `Cabal` but with `Makefile.ghc`. This looks all pretty weird. I'd prefer to have a static `ganeti.cabal` file. This would allow for better structuring of the package description. E.g. `htest` could be a `Test-Suite` instead of a plain `Executable`. The test modules could be part of the Test-Suite and would be hidden from the public Ganeti Haskell API.
I guess, constructing `ganeti.cabal` by a script intended to reduce redundancies that are in the final `ganeti.cabal` file. But the redundancies are acceptable. If not, we have common stanzas since Cabal-2.2. [1]
More advanced would be to generate `ganeti.cabal` from a Yaml file using `hpack`. This is an established solution [2], supported by tools like `stack` and `cabal2nix`. I guess, `cabal/ganeti.template.cabal` predates both Cabal-2.2 and `hpack`.

[1] https://cabal.readthedocs.io/en/stable/cabal-package-description-file.html#common-stanzas
[2] https://packages.debian.org/bookworm/hpack

**Comment by thielema** [comment]
The Python code could go in a Pip package or the Python source could be included in the Cabal package as `Extra-Source-Files`.

**Comment by thielema** [comment]
The Cabal file could look like this:
~~~~
cabal-version:       2.4
name:                ganeti
version:             3.0
homepage:            http://www.ganeti.org
license:             BSD-2-Clause
license-file:        COPYING
author:              Google Inc.
maintainer:          ganeti-devel@googlegroups.com
copyright:           2006-2015 Google Inc.
category:            System
build-type:          Simple
extra-source-files:  README
synopsis:            Cluster-based virtualization management software
description:
  Cluster-based virtualization management software
  .
  See <http://www.ganeti.org>


flag mond
  description: enable the ganeti monitoring daemon
  default:     True

flag metad
  description: enable the ganeti metadata daemon
  default:     True

-- superseeded by --enable-tests
flag htest
  description: enable tests
  default:     True

flag mcio_xformers
  description: use MonadCatchIO-transformers (deprecated)
  default:     True

flag network_bsd
  description: use the split network-bsd package
  default:     False

library
  exposed-modules:
    AutoConf
    Ganeti.BasicTypes
    Ganeti.Codec
    Ganeti.Common
    Ganeti.Compat
    Ganeti.Confd.Client
    Ganeti.Confd.ClientFunctions
    Ganeti.Confd.Server
    Ganeti.Confd.Types
    Ganeti.Confd.Utils
    Ganeti.Config
    Ganeti.ConfigReader
    Ganeti.ConstantUtils
    Ganeti.Constants
    Ganeti.Cpu.LoadParser
    Ganeti.Cpu.Types
    Ganeti.Curl.Internal
    Ganeti.Curl.Multi
    Ganeti.Daemon
    Ganeti.Daemon.Utils
    Ganeti.DataCollectors
    Ganeti.DataCollectors.CLI
    Ganeti.DataCollectors.CPUload
    Ganeti.DataCollectors.Diskstats
    Ganeti.DataCollectors.Drbd
    Ganeti.DataCollectors.InstStatus
    Ganeti.DataCollectors.InstStatusTypes
    Ganeti.DataCollectors.Lv
    Ganeti.DataCollectors.Program
    Ganeti.DataCollectors.Types
    Ganeti.DataCollectors.XenCpuLoad
    Ganeti.Errors
    Ganeti.HTools.AlgorithmParams
    Ganeti.HTools.Backend.IAlloc
    Ganeti.HTools.Backend.Luxi
    Ganeti.HTools.Backend.MonD
    Ganeti.HTools.Backend.Rapi
    Ganeti.HTools.Backend.Simu
    Ganeti.HTools.Backend.Text
    Ganeti.HTools.CLI
    Ganeti.HTools.Cluster
    Ganeti.HTools.Cluster.AllocatePrimitives
    Ganeti.HTools.Cluster.AllocateSecondary
    Ganeti.HTools.Cluster.AllocationSolution
    Ganeti.HTools.Cluster.Evacuate
    Ganeti.HTools.Cluster.Metrics
    Ganeti.HTools.Cluster.Moves
    Ganeti.HTools.Cluster.Utils
    Ganeti.HTools.Container
    Ganeti.HTools.Dedicated
    Ganeti.HTools.ExtLoader
    Ganeti.HTools.GlobalN1
    Ganeti.HTools.Graph
    Ganeti.HTools.Group
    Ganeti.HTools.Instance
    Ganeti.HTools.Loader
    Ganeti.HTools.Nic
    Ganeti.HTools.Node
    Ganeti.HTools.PeerMap
    Ganeti.HTools.Program.Hail
    Ganeti.HTools.Program.Harep
    Ganeti.HTools.Program.Hbal
    Ganeti.HTools.Program.Hcheck
    Ganeti.HTools.Program.Hinfo
    Ganeti.HTools.Program.Hroller
    Ganeti.HTools.Program.Hscan
    Ganeti.HTools.Program.Hspace
    Ganeti.HTools.Program.Hsqueeze
    Ganeti.HTools.Program.Main
    Ganeti.HTools.Tags
    Ganeti.HTools.Tags.Constants
    Ganeti.HTools.Types
    Ganeti.Hash
    Ganeti.Hs2Py.GenConstants
    Ganeti.Hs2Py.GenOpCodes
    Ganeti.Hs2Py.ListConstants
    Ganeti.Hs2Py.OpDoc
    Ganeti.Hypervisor.Xen
    Ganeti.Hypervisor.Xen.Types
    Ganeti.Hypervisor.Xen.XlParser
    Ganeti.JQScheduler
    Ganeti.JQScheduler.Filtering
    Ganeti.JQScheduler.ReasonRateLimiting
    Ganeti.JQScheduler.Types
    Ganeti.JQueue
    Ganeti.JQueue.Lens
    Ganeti.JQueue.Objects
    Ganeti.JSON
    Ganeti.Jobs
    Ganeti.Kvmd
    Ganeti.Lens
    Ganeti.Locking.Allocation
    Ganeti.Locking.Locks
    Ganeti.Locking.Types
    Ganeti.Locking.Waiting
    Ganeti.Logging
    Ganeti.Logging.Lifted
    Ganeti.Logging.WriterLog
    Ganeti.Luxi
    Ganeti.Network
    Ganeti.Objects
    Ganeti.Objects.BitArray
    Ganeti.Objects.Disk
    Ganeti.Objects.Instance
    Ganeti.Objects.Lens
    Ganeti.Objects.Nic
    Ganeti.OpCodes
    Ganeti.OpCodes.Lens
    Ganeti.OpParams
    Ganeti.Parsers
    Ganeti.PartialParams
    Ganeti.Path
    Ganeti.PyValue
    Ganeti.Query.Cluster
    Ganeti.Query.Common
    Ganeti.Query.Exec
    Ganeti.Query.Export
    Ganeti.Query.Filter
    Ganeti.Query.FilterRules
    Ganeti.Query.Group
    Ganeti.Query.Instance
    Ganeti.Query.Job
    Ganeti.Query.Language
    Ganeti.Query.Locks
    Ganeti.Query.Network
    Ganeti.Query.Node
    Ganeti.Query.Query
    Ganeti.Query.Server
    Ganeti.Query.Types
    Ganeti.Rpc
    Ganeti.Runtime
    Ganeti.SlotMap
    Ganeti.Ssconf
    Ganeti.Storage.Diskstats.Parser
    Ganeti.Storage.Diskstats.Types
    Ganeti.Storage.Drbd.Parser
    Ganeti.Storage.Drbd.Types
    Ganeti.Storage.Lvm.LVParser
    Ganeti.Storage.Lvm.Types
    Ganeti.Storage.Utils
    Ganeti.THH
    Ganeti.THH.Compat
    Ganeti.THH.Field
    Ganeti.THH.HsRPC
    Ganeti.THH.PyRPC
    Ganeti.THH.PyType
    Ganeti.THH.RPC
    Ganeti.THH.Types
    Ganeti.Types
    Ganeti.UDSServer
    Ganeti.Utils
    Ganeti.Utils.AsyncWorker
    Ganeti.Utils.Atomic
    Ganeti.Utils.IORef
    Ganeti.Utils.Livelock
    Ganeti.Utils.MVarLock
    Ganeti.Utils.Monad
    Ganeti.Utils.MultiMap
    Ganeti.Utils.Random
    Ganeti.Utils.Statistics
    Ganeti.Utils.UniStd
    Ganeti.Utils.Validate
    Ganeti.VCluster
    Ganeti.Version
    Ganeti.WConfd.Client
    Ganeti.WConfd.ConfigModifications
    Ganeti.WConfd.ConfigState
    Ganeti.WConfd.ConfigVerify
    Ganeti.WConfd.ConfigWriter
    Ganeti.WConfd.Core
    Ganeti.WConfd.DeathDetection
    Ganeti.WConfd.Language
    Ganeti.WConfd.Monad
    Ganeti.WConfd.Persistent
    Ganeti.WConfd.Server
    Ganeti.WConfd.Ssconf
    Ganeti.WConfd.TempRes
  other-extensions:
      TemplateHaskell
  build-depends:
      base                          >= 4.9.0.0
    , array                         >= 0.4.0.0
    , bytestring                    >= 0.9.2.1
    , containers                    >= 0.4.2.1
    , cryptonite                    >= 0.23
    , deepseq                       >= 1.3.0.0
    , directory                     >= 1.1.0.2
    , filepath                      >= 1.3.0.0
    , mtl                           >= 2.1.1
    , old-time                      >= 1.1.0.0
    , pretty                        >= 1.1.1.0
    , process                       >= 1.1.0.1
    , random                        >= 1.0.1.1
    , template-haskell              >= 2.11.0.0
    , text                          >= 0.11.1.13
    , transformers                  >= 0.3.0.0    && < 0.7
    , utf8-string                   >= 0.3.7

    , attoparsec                    >= 0.10.1.1   && < 0.15
    , base64-bytestring             >= 1.0.0.1    && < 1.3
    , case-insensitive              >= 0.4.0.1    && < 1.3
    , curl                          >= 1.3.7      && < 1.4
    , hinotify                      >= 0.3.2      && < 0.5
    , hslogger                      >= 1.1.4      && < 1.4
    , json                          >= 0.5        && < 1.0
    , lens                          >= 3.10       && < 6.0
    , lifted-base                   >= 0.2.0.3    && < 0.3
    , monad-control                 >= 0.3.1.3    && < 1.1
    , parallel                      >= 3.2.0.2    && < 3.3
    , regex-pcre                    >= 0.94.2     && < 0.96
    , transformers-base             >= 0.4.1      && < 0.5
    , unix                          >= 2.5.1.0    && < 2.9
    , zlib                          >= 0.5.3.3    && < 0.8

  if flag(network_bsd)
    build-depends:
        network                     >= 2.9        && < 3.3
      , network-bsd                 >= 2.8        && < 2.9
  else
    build-depends:
        network                     >= 2.3.0.13   && < 2.9

  if flag(mond)
    build-depends:
        PSQueue                       >= 1.1        && < 1.3
      , snap-core                     >= 1.0.0
      , snap-server                   >= 1.0.0
    exposed-modules:
      Ganeti.Monitoring.Server

  if flag(metad)
    build-depends:
        snap-core                     >= 0.8.1
      , snap-server                   >= 0.8.1
    exposed-modules:
      Ganeti.Metad.Config
      Ganeti.Metad.ConfigCore
      Ganeti.Metad.ConfigServer
      Ganeti.Metad.Server
      Ganeti.Metad.Types
      Ganeti.Metad.WebServer


  if flag(mcio_xformers)
    build-depends:
        MonadCatchIO-transformers

  hs-source-dirs:     src
  build-tool-depends: hsc2hs:hsc2hs
  default-language:   Haskell2010
  ghc-options:
    -Wall
    -ddump-splices -ddump-to-file
--    -fwarn-monomorphism-restriction

common app
  hs-source-dirs: apps
  default-language: Haskell2010
  build-depends: ganeti, base

executable htools
  import: app
  main-is: htools.hs

executable hpc-htools
  import: app
  main-is: hpc-htools.hs

executable hpc-mon-collector
  import: app
  main-is: hpc-mon-collector.hs

executable ganeti-kvmd
  import: app
  main-is: ganeti-kvmd.hs

executable ganeti-wconfd
  import: app
  main-is: ganeti-wconfd.hs

executable hconfd
  import: app
  main-is: hconfd.hs

executable hluxid
  import: app
  main-is: hluxid.hs

executable hs2py
  import: app
  main-is: hs2py.hs
  if !flag(metad)
    buildable: False

executable rpc-test
  import: app
  main-is: rpc-test.hs
  build-depends: json

executable ganeti-mond
  import: app
  main-is: ganeti-mond.hs
  if !flag(mond)
    buildable: False

executable ganeti-metad
  import: app
  main-is: ganeti-metad.hs
  if !flag(metad)
    buildable: False

test-suite htest
  import: app
  type: exitcode-stdio-1.0
  hs-source-dirs: test/hs
  main-is: htest.hs
  build-depends:
    HUnit                         >= 1.2.4.2    && < 1.7,
    QuickCheck                    >= 2.8        && < 2.16,
    test-framework                >= 0.6        && < 0.9,
    test-framework-hunit          >= 0.2.7      && < 0.4,
    test-framework-quickcheck2    >= 0.2.12.1   && < 0.4,
    temporary                     >= 1.1.2.3    && < 1.4,
    attoparsec,
    containers,
    text,
    bytestring,
    utf8-string,
    directory,
    process,
    json,
    template-haskell,
    hslogger,
    base64-bytestring,
    filepath,
    lens,
    network,
    unix,
    old-time
  other-modules:
    Test.AutoConf
    Test.Ganeti.Attoparsec
    Test.Ganeti.BasicTypes
    Test.Ganeti.Common
    Test.Ganeti.Confd.Types
    Test.Ganeti.Confd.Utils
    Test.Ganeti.Constants
    Test.Ganeti.Daemon
    Test.Ganeti.Errors
    Test.Ganeti.HTools.Backend.MonD
    Test.Ganeti.HTools.Backend.Simu
    Test.Ganeti.HTools.Backend.Text
    Test.Ganeti.HTools.CLI
    Test.Ganeti.HTools.Cluster
    Test.Ganeti.HTools.Container
    Test.Ganeti.HTools.Graph
    Test.Ganeti.HTools.Instance
    Test.Ganeti.HTools.Loader
    Test.Ganeti.HTools.Node
    Test.Ganeti.HTools.PeerMap
    Test.Ganeti.HTools.Types
    Test.Ganeti.Hypervisor.Xen.XlParser
    Test.Ganeti.JQScheduler
    Test.Ganeti.JQueue
    Test.Ganeti.JQueue.Objects
    Test.Ganeti.JSON
    Test.Ganeti.Jobs
    Test.Ganeti.Kvmd
    Test.Ganeti.Locking.Allocation
    Test.Ganeti.Locking.Locks
    Test.Ganeti.Locking.Waiting
    Test.Ganeti.Luxi
    Test.Ganeti.Network
    Test.Ganeti.Objects
    Test.Ganeti.Objects.BitArray
    Test.Ganeti.OpCodes
    Test.Ganeti.PartialParams
    Test.Ganeti.PyValue
    Test.Ganeti.Query.Aliases
    Test.Ganeti.Query.Filter
    Test.Ganeti.Query.Instance
    Test.Ganeti.Query.Language
    Test.Ganeti.Query.Network
    Test.Ganeti.Query.Query
    Test.Ganeti.Rpc
    Test.Ganeti.Runtime
    Test.Ganeti.SlotMap
    Test.Ganeti.Ssconf
    Test.Ganeti.Storage.Diskstats.Parser
    Test.Ganeti.Storage.Drbd.Parser
    Test.Ganeti.Storage.Drbd.Types
    Test.Ganeti.Storage.Lvm.LVParser
    Test.Ganeti.THH
    Test.Ganeti.THH.Types
    Test.Ganeti.TestCommon
    Test.Ganeti.TestHTools
    Test.Ganeti.TestHelper
    Test.Ganeti.TestImports
    Test.Ganeti.Types
    Test.Ganeti.Utils
    Test.Ganeti.Utils.MultiMap
    Test.Ganeti.Utils.Statistics
    Test.Ganeti.WConfd.Ssconf
    Test.Ganeti.WConfd.TempRes
~~~~

**Comment by apoikos** [comment]
Hi @thielema! With the Haskell code not changing that often, I'm all for simplifying the build system and shipping a static ganeti.cabal file. I played around a bit with this idea back in 2018 (!), using `hpack` and a static `package.yaml`, and getting completely rid of `Makefile.ghc` - I pushed all work-in-progress from back then to the [build-system-overhaul](https://github.com/ganeti/ganeti/tree/build-system-overhaul) branch. The initial motivation there was to support builds using stack, so that we could easily test against newer GHC versions without waiting for the respective Debian builds. Since you are a lot more up-to-date with the latest Haskell trends, are `stack` / `hpack` still a thing?

**Comment by thielema** [comment]
On Thu, 27 Jun 2024, Apollon Oikonomopoulos wrote:

> Hi @thielema! With the Haskell code not changing that often, I'm all for 
> simplifying the build system and shipping a static ganeti.cabal file. I 
> played around a bit with this idea back in 2018 (!), using hpack and a 
> static package.yaml, and getting completely rid of Makefile.ghc - I 
> pushed all work-in-progress from back then to the build-system-overhaul 
> branch.

That's good (old) news. I will inspect that branch.

> The initial motivation there was to support builds using stack, so that 
> we could easily test against newer GHC versions without waiting for the 
> respective Debian builds. Since you are a lot more up-to-date with the 
> latest Haskell trends, are stack / hpack still a thing?

I use neither 'stack' nor 'hpack' for my own development, because I try to 
maximize the range of GHC/base versions my packages are compatible to. 
Stack instead always picks one GHC version and the newest libraries 
available at the time of a new GHC release.

I think 'stack' is good for providing a consistent package set for 
GNU/Linux distributions. You should use the --system-ghc option, though, 
in order to use the GHC version from your distribution. Without that 
option 'stack' downloads a local copy of the GHC version associated with a 
certain Stackage release.

I think, hpack has become less relevant since Cabal got Common stanzas, 
which I use in my proposed ganeti.cabal.

**Comment by thielema** [comment]
On Thu, 27 Jun 2024, Apollon Oikonomopoulos wrote:

> Hi @thielema! With the Haskell code not changing that often, I'm all for 
> simplifying the build system and shipping a static ganeti.cabal file. I 
> played around a bit with this idea back in 2018 (!), using hpack and a 
> static package.yaml, and getting completely rid of Makefile.ghc - I 
> pushed all work-in-progress from back then to the build-system-overhaul 
> branch.

What parts of that branch are still usable today? Why were your changes 
not merged into master?

**Comment by thielema** [comment]
There are so many options of the end result of the build system and even more ways to go there. We could make Python modules extra source files of a Haskell-Cabal project, or Haskell stuff part of a Python-PIP, or we could prepare both Haskell-Cabal and Python-PIP packages and glue them together with some Make/Configure stuff. Automake is intended for writing Makefiles for different Make implementations (make, gnumake). I do not think we still need this, simple Autoconf should suffice. But properly removing Automake stuff might not be simple.

I think, I'd go with Haskell-Cabal + Python-PIP + top-level Makefile to control both components.

Problem is, that Cabal is not a drop-in replacement for the Make stuff. It comes with its own directory structure and configuration files. For instance Cabal writes all object files and executables to directories `dist` or `dist-newstyle`, which is good, but different from what Make currently does. Currently Make puts the object files alongside the source files. Thus it needs to keep track of what are source files and what are generated files in order to do proper `sdist` and `clean`. In Cabal's directory structure `clean` simply means deleting `dist/build`.

We need  a way to install executables to a destination. Cabal usually installs to `$HOME/.cabal/bin` but can be instructed to write to other locations. Makefile has a target for every executable, but for Cabal it would be preferable to build and install all executables at once.

`cabal sdist` generates a source tarball. But that one would miss the Python stuff. So I guess we need to make Cabal and PIP tarballs, append them and add the global control Makefile. Or make the Makefile an Extra-Source-File of the Cabal project.

`Cabal` generally does not write to any other directory than `dist` or `dist-newstyle` in the project directory. This means it also does not write the results of preprocessing to the source directory, but also to `dist`. This is good, but different from what Make currently does. So e.g. AutoConf.hs should be compiled but should not be in the source tarball generated by `cabal sdist`. So the clean way would be to teach Cabal to create AutoConf.hs by running a preprocessor.

As a first quick hack we could generate AutoConf.hs & et.al. by the top-level Makefile and pretend to Cabal that it is a regular source file. But then Cabal would put it into the source tarball and we had to remove it for release, again.

Make has its parallel build feature (option `-j`) and Cabal has its own one (also called `-j`) and GHC has another one (also called `-j` for consistency). GHC's `-j` sets the maximum of modules to be compiled in parallel, Cabal's `-j` limits the maximum number of packages to be build in parallel and Make could perform other stuff in parallel to Cabal builds. This is not good because we actually want a global limit for maximum number of cores to use. This is currently achieved with the Make-only build system, where Make controls the compilation of every Haskell module. GHC and Cabal developers are working on merging parallelization of Haskell compilation, that is, Cabal splits compilation of many packages into smaller tasks at the level of compilation of single modules, and GHC processes them all at once while respecting the maximum of parallelism. This is system not yet ready, and when it will be, it will still not cooperate with Make.


That said, there are some difference between Cabal and Autoconf/Make and I wonder whether I should try to replace Make stuff step by step by Cabal stuff. Or whether we should start a new independent build system and throw away Automake when we got all functionality with the new system. But there are some file names like `Makefile` and `ganeti.cabal` that would be naturally used also be the new system.

So, these are some extended thoughts on the build system. I am uncertain how to proceed.

**Comment by thielema** [comment]
I think I will try these two next steps:

* Replace generated `ganeti.cabal` by static one. I do not see why it was dynamically generated, at all. Automake already converts Configure flags to Cabal flags. Currently `ganeti.cabal` seems to be used only for generating `cabal_macros.h` and `ganeti.depsflags`. But then dependency analysis and compilation is performed manually via Make and direct calls to GHC.

* Compile with Cabal using static `ganeti.cabal`, but leave cleaning and source tarball generation to Automake, for now. Cabal generates `cabal_macros.h` and instead of writing the package set to `ganeti.depsflags` it passes according options to GHC directly. This way we can get rid of `cabal/CabalDependenciesMacros.hs` which contains many switches for different versions of Cabal, which is cumbersome and prone to fail.

By the way, do we have CI actions that builds Ganeti from a source tarball? It would be important to see whether generated source tarballs are complete.

**Comment by saschalucas** [comment]
Thanks @thielema for your explanation in https://github.com/ganeti/ganeti/issues/1747#issuecomment-2254104386. You already answered your self with:
  
> I think I will try these two next steps:
> 
>     * Replace generated `ganeti.cabal` by static one. ...
>     * Compile with Cabal using static `ganeti.cabal`, but leave cleaning and source tarball generation to Automake. ...

That sounds reasonable. And you already noticed in #1768 that both steps need to be done at once.

Since we can not replace the entire autotools build system at once for both python and Haskell, the hope is to switch just the Haskell compilation part to `Cabal` and integrate this as a secondary build system into `Makefile.am` ???

> By the way, do we have CI actions that builds Ganeti from a source tarball? It would be important to see whether generated source tarballs are complete.

Actually no. I think there are some targets in `Makefile.am` that have this indentation, i.e. `distrebuildcheck`. However I also think, that they are incomplete, because `./configure` inside this target does not enable all flags, i.e. building all Haskell Daemons like `--enable-monitoring` and `--enable-metadata`. OTH this two Daemons are probably not used by anybody or even not functional at all.

If there is still the need to have this in CI, please specify what make target you have in mind, I can then try to add it.
